### ISSUE TITLE
pam auth_type not supported in hba.conf in version 1.8.1

### ISSUE THREAD

**Issue by svb007** [pgbouncer/pgbouncer#276]
This feature seems to be very easy to enable with the following patch:

diff -urN pgbouncer-1.8.1-orig/src/hba.c pgbouncer-1.8.1/src/hba.c
--- pgbouncer-1.8.1-orig/src/hba.c      2017-12-18 16:03:18.000000000 +0000
+++ pgbouncer-1.8.1/src/hba.c   2018-01-23 22:12:05.163761334 +0000
@@ -589,6 +589,8 @@
                rule->rule_method = AUTH_PEER;
        } else if (eat_kw(tp, "cert")) {
                rule->rule_method = AUTH_CERT;
+       } else if (eat_kw(tp, "pam")) {
+               rule->rule_method = AUTH_PAM;
        } else {
                log_warning("hba line %d: unsupported method: buf=%s", linenr, tp->buf);
                goto failed;

**Comment by svb007** [comment]
It seems the above patch causes an issue with pgbouncer not able to recover from certain connection problems that trigger the "pgbouncer cannot connect to server" message, because the pool count gets messed up. Below is a patch I included to try and work around this. I hope one of the developers can find a way of implementing auth_type pam in hba.conf without these issues

[cannot_connect_to_server.patch.txt](https://github.com/pgbouncer/pgbouncer/files/1721746/cannot_connect_to_server.patch.txt)

**Comment by svb007** [comment]
It seems that the problem reported in https://github.com/pgbouncer/pgbouncer/issues/285, confirms that it wasn't the pam_hba.patch that caused this behaviour, but that it is a bug in the current version of pgbouncer

**Comment by svb007** [comment]
I agree with that, I think you just need to fix 3 (It references the "now" variable which I did define in 2), and if you don't mind, please post your PR, and then link me to it so I can see how it's done, then I will post 4 as a patch. 1 still needs some work, as I believe it should be added as part of the --with-pam option, so I agree we shouldn't include that one.

**Comment by achix** [comment]
ok.

**Comment by achix** [comment]
Hello I submitted all in one : 
https://github.com/pgbouncer/pgbouncer/pull/361

it solves also the pgbouncer crash when strlen(username) > 64.

phewww first PR ever :)

**Comment by achix** [comment]
yep! that's a pain keeping the patches, at least we can keep this as a
reference in case the PR is rejected :)

**Comment by svb007** [comment]
I have been running with this patch in place for quite some time now, and I can now confirm that I have not experienced any issues with it. The only issue I had was fixed by another commit (https://github.com/pgbouncer/pgbouncer/commit/b30b49f16d718339f0f7ffee408c48b376c0eb81), and was therefore unrelated to this patch. I think including this by default (if compiled with --with-pam) is safe

**Comment by achix** [comment]
@svb007 why don't you submit a PR ?
Will you attend next pgconf.eu ?

**Comment by svb007** [comment]
Well I tried to today : https://github.com/svb007/pgbouncer/commit/db88677f3882b22d208fcce41cd33f4cc6d2ff68

Not sure if was done correctly.
I'm in New Zealand, so probably won't make pgconf.eu. Would have loved to though

**Comment by svb007** [comment]
Ah,  the PR is here : https://github.com/pgbouncer/pgbouncer/pull/422
